docs/concepts/experimental.md

```diff
@@ -71,10 +71,6 @@ class User(BaseModel):
         ),
     ]
     friends: Annotated[list[User], validate_as(...).len(0, 100)]  # (6)!
-    family: Annotated[  # (7)!
-        list[User],
-        validate_as_deferred(lambda: list[User]).transform(lambda x: x[1:]),
-    ]
```

Comment:
Copy pasting what I added on Slack so that it's not lost:

the TL;DR is using lambdas with references to other symbols in annotations opens the door to a lot of weird behaviors.

---

This is the (simplified) test failing on the PR:

```python
from typing_extensions import Annotated

from pydantic import BaseModel
from pydantic.experimental.pipeline import validate_as_deferred

class User(BaseModel):
    family: 'Annotated[list[User], validate_as_deferred(lambda: list[User])]'

# The `family` annotation is successfully evaluated, but at some point during
# core schema generation, the pipeline API logic is triggered and when the lambda
# gets called, we end up with:
# NameError: name 'User' is not defined
```

On `main`, when we evaluate (read: Python will call eval) the annotation for family, we use the following as globals:
`{'User': __main__.User, 'Annotated': ..., 'BaseModel': ..., ...}`
and locals are empty.

On this PR, we cleaned up how globals and locals were mixed up before. This means that we now use the following as globals:
`{'Annotated': ..., 'BaseModel': ..., ...}`
and locals:
`{'User': __main__.User, ...}`

And the issue comes from what could be considered as a limitation of `eval`. Consider this example:

```python
def func():
    A = int

    works = lambda: list[A]
    fails = eval('lambda: list[A]', globals(), locals())

    works()
    # list[int]
    fails()
    # NameError: A is not defined.
```

The `eval` limitation is that it does not have access to the non-locals of the `lambda` environment (which is a new scope, like with a `def` statement). Even though `A` is present in locals, it won't be used to resolve `A` and so `eval` will look up in the `globals` instead (that's why it works on `main` because `User` was added in `globals` for the `eval` call).

This limitation is documented in this open CPython [PR](https://github.com/python/cpython/pull/25039/files#diff-6a7a07ac473fdd76734669b1b70626ad2176011129902f6add017810f54d0439R551-R555).